### PR TITLE
[SYCL] Implement sycl_ext_oneapi_if_device

### PR DESCRIPTION
**Pull request by aelovikov-intel**
Specification PR is at https://github.com/intel/llvm/pull/8917.

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
So that we could start removing `__SYCL_DEVICE_ONLY__` from the SYCL headers.

File: sycl/include/sycl/ext/oneapi/experimental/if_device.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#pragma once
```
Comment (bader) [inline sycl/include/sycl/ext/oneapi/experimental/if_device.hpp:1]: Please, add a copyright header.

File: sycl/test-e2e/Basic/if_device.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <sycl/sycl.hpp>
```
Comment (bader) [inline sycl/test-e2e/Basic/if_device.cpp:4]: Please, include minimal required set of headers: queue, buffer, accessor and `<sycl/ext/oneapi/experimental/if_device.hpp>`.

Comment (aelovikov-intel) [reply sycl/test-e2e/Basic/if_device.cpp:4]: I strongly disagree. The right process to do that is to update the spec/create extension, allocate resources to the activity, cleanup the headers and create automation/checks to automatically cleanup includes and only after that can we start requesting such changes in arbitrary PRs. Until then we have to treat end-to-end tests same way as examples and write them how customers would write their code according to the specification.

**Review by sergey-semenov — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
Failures:

```
# Intel:
Failed Tests (1):
SYCL :: ESIMD/accessor_local.cpp

# HIP:
Failed Tests (4):
  SYCL :: USM/memops2d/copy2d_device_to_host.cpp
  SYCL :: USM/memops2d/copy2d_host_to_device.cpp
  SYCL :: USM/memops2d/memcpy2d_device_to_host.cpp
  SYCL :: USM/memops2d/memcpy2d_host_to_device.cpp

# CUDA:
Unexpectedly Passed Tests (1):
  SYCL :: DeviceArchitecture/device_architecture_on_host.cpp
```

covered by 

https://github.com/intel/llvm/issues/10138
https://github.com/intel/llvm/issues/10157

and the last is mentioned/awaiting followup in https://github.com/intel/llvm/pull/10173#issuecomment-1618317654.

**Review by gmlueck — APPROVED** [review]

**Comment by bader** [comment]
@aelovikov-intel, this seems to be ready to merge. What is holding this back?

**Comment by gmlueck** [comment]
Actually, this should not be merged yet.  There was a recent change in the API, and the `fn` parameter is now passed via perfect forwarding (i.e. as `T &&Fn`).  See #8917.

**Pull request closed without merging**